### PR TITLE
Fix some to some.

### PR DESCRIPTION
**Pull request by luca-heltai**
When calling `Utilities::MPI::some_to_some`, if the map contained our own mpi process, an assert would be thrown stating that we cannot communicate with our own process.

This PR modifies this behaviour, and makes sure that anything that we try to send to our selves is actually simply copied in the output map, simplifying the logic behind many algorithms.

### CODE REVIEW

**Comment by luca-heltai** [comment]
@blaisb , FYI. :)

**Comment by blaisb** [comment]
Awesome. This will be very useful !

File: include/deal.II/base/mpi.h [hunk 1]
```diff
@@ -1481,21 +1481,24 @@ namespace Utilities
...
-        unsigned int i = 0;
-        for (const auto &m : objects_to_send)
-          send_to[i++] = m.first;
-      }
-      AssertDimension(send_to.size(), objects_to_send.size());
+      std::vector<unsigned int> send_to;
+      send_to.reserve(objects_to_send.size());
+      for (const auto &m : objects_to_send)
+        if (m.first != my_proc)
+          send_to.emplace_back(m.first);
```
Comment (peterrum) [inline include/deal.II/base/mpi.h]:
```suggestion
      for (const auto &m : objects_to_send)
        if (m.first==my_proc)
          received_objects[my_proc] = m.second;
        else
          send_to.emplace_back(m.first);
```

**Comment by luca-heltai** [comment]
The failed test is unrelated to this pull request.

```

99% tests passed, 1 tests failed out of 605



Total Test time (real) = 1042.97 sec



The following tests FAILED:

	4111 - multigrid/transfer_04b.mpirun=2.debug (Timeout)

Errors while running CTest

script returned exit code 8

```

File: include/deal.II/base/mpi.h [hunk 2]
```diff
@@ -1481,21 +1481,28 @@ namespace Utilities
...
+          received_objects[my_proc] = m.second;
+        else
+          send_to.emplace_back(m.first);
+
+      // Shortcut, when running in serial
+      if (send_to.size() == 0)
+        return received_objects;
 
       const auto receive_from =
         Utilities::MPI::compute_point_to_point_communication_pattern(comm,
```
Comment (masterleinad) [inline include/deal.II/base/mpi.h]: Only the number of messages to receive is required from this function later on (l.1544). Hence, `Utilities::MPI::compute_n_point_to_point_communications` should suffice here.

**Comment by luca-heltai** [comment]
> Looks good to me, but I couldn 't find the discussion you were referring to in #9069 (comment).

I could not find it either. Maybe I'm confusing this with other stuff we did in the same period... My bad.

File: include/deal.II/base/mpi.h [hunk 3]
```diff
@@ -1481,25 +1481,31 @@ namespace Utilities
...
+      send_to.reserve(objects_to_send.size());
+      for (const auto &m : objects_to_send)
+        if (m.first == my_proc)
+          received_objects[my_proc] = m.second;
+        else
+          send_to.emplace_back(m.first);
+
+      // Shortcut, when running in serial
+      if (send_to.size() == 0)
+        return received_objects;
```
Comment (peterrum) [inline include/deal.II/base/mpi.h]: This is the reason why  `multigrid/transfer_04b.mpirun=2.debug` fails. `compute_n_point_to_point_communications` is a  global communication step!

Comment (luca-heltai) [reply include/deal.II/base/mpi.h]: I did not know we used some_to_some in multigrid transfer... :) I'm on it.

File: include/deal.II/base/mpi.h [hunk 4]
```diff
@@ -1481,25 +1481,31 @@ namespace Utilities
...
+          send_to.emplace_back(m.first);
+
+      const auto n_point_point_communications =
+        Utilities::MPI::compute_n_point_to_point_communications(comm, send_to);
 
-      const auto receive_from =
-        Utilities::MPI::compute_point_to_point_communication_pattern(comm,
-                                                                     send_to);
+      // Shortcut, when we have nothing to send or nothing to receive
+      if (send_to.size() == 0 && n_point_point_communications == 0)
```
Comment (masterleinad) [inline include/deal.II/base/mpi.h]:
```suggestion
      if (send_to.size() == 0 || n_point_point_communications == 0)
```
?

Comment (luca-heltai) [reply include/deal.II/base/mpi.h]: Why? At least one of the two blocks below would not be visited with ||, leading to a deadlock.

Comment (masterleinad) [reply include/deal.II/base/mpi.h]: Yes, makes sense. The sentence above confused me. I guess something like
"Shortcut, when we have neither anything to send nor anything to receive" or "Shortcut, when we have nothing to send and nothing to receive"

It's not important to me, though.

Comment (masterleinad) [reply include/deal.II/base/mpi.h]: Thanks!

**Pull request merged**